Stop NamedTemporaryFile tampering

adamnovak opened the pull request:
This builds on #2607 and is an attempt to get it and all the other PRs failing on the autoscaling Gitlab runner to pass.

adamnovak commented:
OK, here's the real problem:

```
WARNING:toil.leader:Job failed with exit value 138: 'setup' s/i/jobYOZfTG
WARNING:toil.leader:The job seems to have left a log file, indicating failure: 'setup' s/i/jobYOZfTG
WARNING:toil.leader:s/i/jobYOZfTG    INFO:toil.worker:---TOIL WORKER OUTPUT LOG---
WARNING:toil.leader:s/i/jobYOZfTG    INFO:toil:Running Toil version 3.20.0a1-c0a1d2d5cf8b0f285705e65da2dc34e102301773.
WARNING:toil.leader:s/i/jobYOZfTG    CRITICAL:toil.jobStores.fileJobStore:jobStoreFilePath: /tmp/toil-test-toil.test.sort.sortTest.SortTest-testFileSingleNonCaching-jobstore-g_t4rU/tmp/s/i/jobYOZfTG/g/tmpjU_X2K-_serialiseJob-stream True
WARNING:toil.leader:s/i/jobYOZfTG    CRITICAL:toil.jobStores.fileJobStore:localFilePath: /builds/databiosphere/toil/tmp/tmpiPK5BL/job False
WARNING:toil.leader:s/i/jobYOZfTG    Traceback (most recent call last):
WARNING:toil.leader:s/i/jobYOZfTG      File "/builds/databiosphere/toil/src/toil/worker.py", line 316, in workerScript
WARNING:toil.leader:s/i/jobYOZfTG        job = Job._loadJob(jobGraph.command, jobStore)
WARNING:toil.leader:s/i/jobYOZfTG      File "/builds/databiosphere/toil/src/toil/job.py", line 930, in _loadJob
WARNING:toil.leader:s/i/jobYOZfTG        jobStore.readFile(pickleFile, filename)
WARNING:toil.leader:s/i/jobYOZfTG      File "/builds/databiosphere/toil/src/toil/jobStores/fileJobStore.py", line 390, in readFile
WARNING:toil.leader:s/i/jobYOZfTG        os.link(jobStoreFilePath, localFilePath)
WARNING:toil.leader:s/i/jobYOZfTG    OSError: [Errno 18] Invalid cross-device link
WARNING:toil.leader:s/i/jobYOZfTG    ERROR:toil.worker:Exiting the worker because of a failed job on host runner-P2a5xjE1-project-3-concurrent-0
WARNING:toil.leader:s/i/jobYOZfTG    WARNING:toil.jobGraph:Due to failure we are reducing the remaining retry count of job 'setup' s/i/jobYOZfTG with ID s/i/jobYOZfTG to 0
```
That's here:

https://github.com/DataBiosphere/toil/blob/d1d102e058b18a1cc100f40a095b952fc7e03a7e/src/toil/jobStores/fileJobStore.py#L390

We're checking to avoid cross-device links, but I think containerization is causing this check to come out wrong. You can get an invalid cross-device link error when trying to hardlink between different Docker mounts, even though they report as being the same device:

```
docker run -it --rm -v /tmp:/tmp1 -v /tmp/another:/tmp2 ubuntu:18.04

root@5f3f66e8003e:/# touch /tmp2/lol
root@5f3f66e8003e:/# stat /tmp2/lol
  File: /tmp2/lol
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: fd01h/64769d	Inode: 4456486     Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2019-04-19 17:04:46.810644179 +0000
Modify: 2019-04-19 17:04:46.810644179 +0000
Change: 2019-04-19 17:04:46.810644179 +0000
 Birth: -
root@5f3f66e8003e:/# touch /tmp1/aaa
root@5f3f66e8003e:/# stat /tmp1/aaa
  File: /tmp1/aaa
  Size: 0         	Blocks: 0          IO Block: 4096   regular empty file
Device: fd01h/64769d	Inode: 4456590     Links: 1
Access: (0644/-rw-r--r--)  Uid: (    0/    root)   Gid: (    0/    root)
Access: 2019-04-19 17:04:55.506306289 +0000
Modify: 2019-04-19 17:04:55.506306289 +0000
Change: 2019-04-19 17:04:55.506306289 +0000
 Birth: -
root@5f3f66e8003e:/# ln /tmp1/aaa /tmp2/
ln: failed to create hard link '/tmp2/aaa' => '/tmp1/aaa': Invalid cross-device link
```

We should adopt Python's forgiveness over permission principle and just always try to do the hardlink, detect when it fails due to a cross-device link error, and retry with a symlink or copy.

adamnovak commented:
My altered behavior in the FileJobStore is now upsetting the CachingFileStore, because it tries to predict the behavior of the FileJobStore and guess how many links it ought to have to files in different cases.

adamnovak commented:
I *think* this is going to pass tests, so it should be ready for review.

mr-c commented on src/toil/fileStore.py:
```diff
@@ -703,6 +707,8 @@ def readGlobalFile(self, fileStoreID, userPath=None, cache=True, mutable=False,
                         if os.path.exists('/.'.join(os.path.split(cachedFileName))):
                             os.rename('/.'.join(os.path.split(cachedFileName)), cachedFileName)
+                            # If this is not true we get into trouble in our internal reference counting.
+                            assert(os.stat(cachedFileName).st_nlink == self.nlinkThreshold)
```
assert is removed with compiling to optimised byte code (python -o producing *.pyo files). Maybe convert into an exception?

adamnovak replied:
I think this is properly an assert; if we're right in our reasoning about how the code functions (i.e. the FileStore and the FileJobStore agree about link counts), it should never be hit. It is safe to skip if you are willing to sacrifice debuggability for speed using `-o`.

mr-c replied:
Okay, and all the other asserts?

adamnovak commented:
I think they have the same semantics.

On 4/26/19, Michael R. Crusoe <notifications@github.com> wrote:
> mr-c commented on this pull request.
>
>
>
>> @@ -703,6 +707,8 @@ def readGlobalFile(self, fileStoreID, userPath=None,
>> cache=True, mutable=False,
>                          # recording it in the cache lock file) if
> possible.
>                          if
> os.path.exists('/.'.join(os.path.split(cachedFileName))):
>
> os.rename('/.'.join(os.path.split(cachedFileName)), cachedFileName)
> +                            # If this is not true we get into trouble in
> our internal reference counting.
> +                            assert(os.stat(cachedFileName).st_nlink ==
> self.nlinkThreshold)
>
> Okay, and all the other asserts?
>
> --
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub:
> https://github.com/DataBiosphere/toil/pull/2638#discussion_r279045689
